library/oid.c

```diff
@@ -861,4 +895,168 @@ int mbedtls_oid_get_numeric_string(char *buf, size_t size,
+                                         unsigned char *bound,
+                                         unsigned int value)
+{
+    size_t num_bytes = oid_subidentifier_num_bytes(value);
+    if ((size_t) (bound - *p) < num_bytes) {
+        return MBEDTLS_ERR_OID_BUF_TOO_SMALL;
+    }
+    (*p)[num_bytes - 1] = (unsigned char) (value & 0x7f);
+    value >>= 7;
+
+    for (size_t i = 2; i <= num_bytes; i++) {
+        (*p)[num_bytes - i] = 0x80 | (unsigned char) (value & 0x7f);
+        value >>= 7;
+    }
+    *p += num_bytes;
+
+    return 0;
+}
+
+/* Return the OID for the given x.y.z.... style numeric string  */
+int mbedtls_oid_from_numeric_string(mbedtls_asn1_buf *oid,
+                                    const char *buf, size_t size)
+{
+    int ret = MBEDTLS_ERR_ASN1_INVALID_DATA;
+
+    const char *str_ptr = buf;
+    const char *str_bound = buf + size;
+    int val = 0;
+    size_t encoded_len;
+
+    int component1 = oid_parse_number(&str_ptr, str_bound);
+    if (component1 < 0) {
+        return component1;
+    }
+    if (component1 > 2) {
+        /* First component can't be > 2 */
+        return MBEDTLS_ERR_ASN1_INVALID_DATA;
+    }
+    if (str_ptr >= str_bound || *str_ptr != '.') {
+        return MBEDTLS_ERR_ASN1_INVALID_DATA;
+    }
+    str_ptr++;
+
+    int component2 = oid_parse_number(&str_ptr, str_bound);
+    if (component2 < 0) {
+        return component2;
+    }
+    if ((component1 < 2) && (component2 > 38)) {
+        /* Root nodes 0 and 1 may have up to 39 children */
+        return MBEDTLS_ERR_ASN1_INVALID_DATA;
+    }
+    if (str_ptr < str_bound && *str_ptr != '\0') {
+        if (*str_ptr == '.') {
+            str_ptr++;
+        } else {
+            return MBEDTLS_ERR_ASN1_INVALID_DATA;
+        }
+    }
+
+    if ((UINT_MAX - (unsigned int) component2) <=
+        ((unsigned int) component1 * 40)) {
+        return MBEDTLS_ERR_OID_BUF_TOO_SMALL;
+    }
+    encoded_len = oid_subidentifier_num_bytes(((unsigned int) component1 * 40)
+                                              + (unsigned int) component2);
+
+    while (str_ptr < str_bound && *str_ptr != '\0') {
+        val = oid_parse_number(&str_ptr, str_bound);
+        if (val < 0) {
+            return val;
+        }
+        if (str_ptr < str_bound && *str_ptr != '\0') {
+            if (*str_ptr == '.') {
+                str_ptr++;
+            } else {
+                return MBEDTLS_ERR_ASN1_INVALID_DATA;
+            }
+        }
+
+        size_t num_bytes = oid_subidentifier_num_bytes(val);
+        if ((SIZE_MAX - encoded_len) <= num_bytes) {
+            return MBEDTLS_ERR_OID_BUF_TOO_SMALL;
+        }
+        encoded_len += num_bytes;
+    }
+
+    oid->p = mbedtls_calloc(encoded_len, sizeof(unsigned char));
+    if (oid->p == NULL) {
+        return MBEDTLS_ERR_ASN1_ALLOC_FAILED;
+    }
+    oid->len = encoded_len;
+
+    /* Now that we've allocated the buffer, go back to the start and encode */
```

Comment:
I'm not sure if this would help much. We don't know how many values are in the OID without parsing it, so we'd have to do an approximation like assume the string has as many numbers as possible e.g.:
```c
max_possible_subidentifiers = (strlen(oid_str) / 2) + 1
```
This would feel a bit unwieldy (although admittedly the current solution is not all that neat).

It would also make it more difficult to refactor the function to not use malloc if that's something we want to do in future.

What do you think? I'm happy to make this transformation if it is best.

Another option would be to factor out the repeated code into a single function that either:
* Calculates a length, if no buffer is given
* Encodes into a buffer, if a buffer is given

That would get rid of the slightly ugly code duplication. It would look something like this:
```c
ssize_t encoded_len = oid_parse_str_into(NULL, 0, str_ptr, str_bound);
if (encoded_len < 0) {
    return encoded_len;
}

oid->buf = mbedtls_calloc(encoded_len, 1);
if (oid->buf == NULL) {
    return MBEDTLS_ASN1_ERR_ALLOC_FAILED;
}
oid->len = encoded_len;

encoded_len = oid_parse_str_into(oid->buf, oid->len, str_ptr, str_bound);
if (encoded_len < 0) {
    return encoded_len;
}
```
Would this be preferable?

Reply:
Ah okay I see. It will probably be more like a 75% overhead - technically we support subidentifiers up to size 2^32 but in practice most fit in 1 byte, so we're talking about 4x. However, OIDs are (relatively) small as far as it goes, so maybe it would be fine.

Apparently the longest OID that currently exists has 34 arcs ([source](http://oid-info.com/faq.htm#size-limitations)): 
```
1.3.6.1.4.1.1248.1.1.2.1.3.21.69.112.115.111.110.32.83.116.121.108.117.115.32.80.114.111.32.52.57.48.48
```
So that's 33 * 4 = 132 bytes (the first 2 subidentifiers are combined into 1) for an OID that would fit into 34 bytes. That seems reasonably small in comparison to our existing code, especially given that this is the most extreme case.

Reply:
Apologies for the long delay on this. I have rewritten the code to use the allocate-reallocate strategy. Given the 75% likely overhead, worst-case allocation with no reallocation seems like a bad idea for memory usage.

My new changes should address this thread, in any case.